Calculator.Deviationlark/CalculatorLibrary/CalculatorLibrary.cs

```diff
@@ -0,0 +1,150 @@
+            writer.WritePropertyName("Operation");
+
+            // Use a switch statement to do the math.
+            switch (op)
+            {
+                case "a":
+                    result = num1 + num2;
+                    History(num1, num2, result, op);
+                    writer.WriteValue("Add");
+                    break;
+                case "s":
+                    result = num1 - num2;
+                    writer.WriteValue("Subtract");
+                    History(num1, num2, result, op);
+                    break;
+                case "m":
+                    result = num1 * num2;
+                    writer.WriteValue("Multiply");
+                    History(num1, num2, result, op);
+                    break;
+                case "d":
+                    // Ask the user to enter a non-zero divisor.
+                    if (num2 != 0)
+                    {
+                        result = num1 / num2;
+                    }
+                    writer.WriteValue("Divide");
+                    History(num1, num2, result, op);
+                    break;
+                // Return text for an incorrect option entry.
+                default:
+                    Console.WriteLine("Thats not an operation.");
+                    writer.WriteValue("Invalid operation");
+                    Console.ReadLine();
+                    break;
+            }
+            writer.WritePropertyName("Result");
+            writer.WriteValue(result);
+            writer.WriteEndObject();
+
+            return result;
+        }
+
+        public void Finish()
+        {
+            writer.WriteEndArray();
+            writer.WriteEndObject();
+            writer.Close();
+        }
+
+        public void History(double num1 = 0, double num2 = 0, double result = 0, string op = "")
+        {
+            op = Operator(op);
+            calculations.Add($"{count + 1}. {num1} {op} {num2} = {result}");
+            results.Add(result);
+            count++;
+        }
+
+        public double ShowHistory()
+        {
+            Console.Clear();
+            double num1 = 0;
+            string? result = "";
+            foreach (string calculation in calculations)
+            {
+                Console.WriteLine(calculation);
+            }
+            while (string.IsNullOrEmpty(result))
+            {
+
+                Console.WriteLine("u. Use previous result");
+                Console.WriteLine("d. Delete history");
+                Console.WriteLine("Enter any other key to do another calculation.");
+                result = Console.ReadLine();
+            }
+            if (result == "u")
+            {
+                Console.WriteLine("Type the number of the result you would like to use: ");
+                result = Console.ReadLine();
+                if (int.TryParse(result, out _) && Convert.ToDouble(result) > 0)
+                    num1 = Convert.ToDouble(results[(Convert.ToInt32(result) - 1)]);
```

Comment:
Well done with the project and good job going for the challenges.

Since you went above and beyond what you needed to do here's one problem you could try to tackle on your own time.

If i try to use the history result for my next calculation and insert a number that is not in the list, the program will close with an exception. As you can see here

![image](https://github.com/TheCSharpAcademy/CodeReviews.Console.Calculator/assets/12613508/6353789b-7e11-4f98-8c15-da3c798738bc)
That said, the project meets the requirements, so i will approve it, good job!

Reply:
Went unnoticed, thanks for the review. I fixed it.